src/main/golem/util/validation/dimensions.kt

```diff
@@ -0,0 +1,256 @@
+@file:JvmName("DimensionValidation")
+
+package golem.util.validation
+import golem.validateMatrices
+import golem.matrix.Matrix
+import kotlin.collections.MutableMap
+import kotlin.collections.mutableMapOf
+import kotlin.collections.MutableSet
+import kotlin.collections.mutableSetOf
+import java.util.HashSet
+import java.io.ByteArrayOutputStream
+import java.io.PrintStream
+
+private val ValidationContext.meta : DimensionValidator get() {
+    return this.metadata("dim") {
+        val instance = DimensionValidator()
+        addValidator(instance)
+        instance
+    }
+}
+
+private class DimensionValidator: Validator {
+    private val possibleValues = mutableMapOf<Char, MutableSet<Int>>()
```

Comment:
Docs on these might be useful-- its not immediately clear what a String-> String,String called "declared" is for example.